proto/workload.proto

```diff
@@ -103,9 +101,20 @@ message Workload {
   // Address represents the IPv4/IPv6 address for the workload.
   // This should be globally unique.
   // This should not have a port number.
-  // Each workload must have at least either an address or hostname; not both.
+  // Each workload must have either an address or async_hostname; not both.
   repeated bytes addresses = 3;
 
+  // The hostname for the workload to be resolved asynchronously by the ztunnel.
+  //
+  // If the resolved DNS query has several endpoints, any subsequent requests will
+  // be load balanced between these endpoints (i.e. envoy STRICT_DNS cluster)
+  // TODO: add support for another load-balancing mode where each connection sends
+  // to the first endpoint resolved by its own DNS query for the lifetime of the
+  // connection in the connection pool (i.e. envoy LOGICAL_DNS cluster)
+  //
+  // Each workload must have either an address or async_hostname; not both.
+  string async_hostname = 23;
```

Comment:
I definitely do not think we should go forward with this. What is the purpose of this again? It's not clear how this is different than `hostname`. I know fundamentally one is supposed to come from WLE and one from pod, but nothing about this naming makes that clear.

Could we instead make `hostname` and array?

@howardjohn wdyt?

Reply:
Am I oversimplifying this or do we just need to do:

```
if workload.ip != null {
 use ip 
} else if workload.hostname != null {
 use hostname
}
```

?
Not sure why we need a new field for that.

A headless pod will have a hostname + IP, so we don't need to use the hostname on the request path (i mean, we could --it would just resolve to the IP though so its moot)

Reply:
I think that is over-simplifying. If we want to keep service entry + workload entry semantics the same as in classic istio then we need async DNS resolution for SE/WE hostnames. Today's behavior:

https://github.com/istio/istio/blob/30b23bab0017d3be258d79230ec05871afcad144/pilot/pkg/serviceregistry/serviceentry/conversion.go#L330 for service entry resolution

https://github.com/istio/istio/blob/30b23bab0017d3be258d79230ec05871afcad144/pilot/pkg/networking/core/v1alpha3/cluster.go#L671-L674 for how that converts to envoy clusters

These envoy clusters from the envoy docs https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/service_discovery#arch-overview-service-discovery-types-strict-dns

> When using strict DNS service discovery, Envoy will continuously and asynchronously resolve the specified DNS targets. Each returned IP address in the DNS result will be considered an explicit host in the upstream cluster. This means that if the query returns three IP addresses, Envoy will assume the cluster has three hosts, and all three should be load balanced to. If a host is removed from the result Envoy assumes it no longer exists and will drain traffic from any existing connection pools. Consequently, if a successful DNS resolution returns 0 hosts, Envoy will assume that the cluster does not have any hosts. ***Note that Envoy never synchronously resolves DNS in the forwarding path. At the expense of eventual consistency, there is never a worry of blocking on a long running DNS query.***

SE has the resolution field (not currently supported) and WE addresses can be hostnames. Both need to be looked at more closely, I'll look more tomorrow. Parity with current features is intent of this PR though

Reply:
comment here added the context i was missing https://github.com/istio/ztunnel/pull/612#issuecomment-1642303664. opted to implement on demand rather than copy envoy's impl